File: src/Object/Object.cpp
#include "Object.hpp"

unsigned short getVertexFeatureSize(VertexFeature feature) {
    switch(feature) {
        case VertexFeature::Position:
            return 3;
        case VertexFeature::Normal:
            return 3;
        case VertexFeature::Color:
            return 3;
        case VertexFeature::UV:
            return 2;
    }

    return 0;
}

Object::Object(
    const std::vector<float> vertices,
    const std::vector<unsigned int> indices,
    const std::vector<VertexFeature> features
) {
    init(vertices, indices, features);
}

Object::Object(
    const std::string& objFilePath,
    const std::vector<VertexFeature> features,
    float scaleOnLoad
) {
    ObjLoader loader(objFilePath);

    std::vector<float> vertices;
    std::vector<unsigned int> indices;

    bool positionEnabled = std::find(features.begin(), features.end(), VertexFeature::Position) != features.end();
    bool normalEnabled = std::find(features.begin(), features.end(), VertexFeature::Normal) != features.end();
    bool colorEnabled = std::find(features.begin(), features.end(), VertexFeature::Color) != features.end();
    bool uvEnabled = std::find(features.begin(), features.end(), VertexFeature::UV) != features.end();

    for(struct ObjFace face : loader.faces) {
        for(struct ObjVertex vertex : face.vertices) {
            if(positionEnabled) {
                vertices.push_back(scaleOnLoad * loader.vertices[vertex.vertexIndex].x);
                vertices.push_back(scaleOnLoad * loader.vertices[vertex.vertexIndex].y);
                vertices.push_back(scaleOnLoad * loader.vertices[vertex.vertexIndex].z);
            }

            if(normalEnabled) {
                vertices.push_back(loader.normals[vertex.normalIndex].x);
                vertices.push_back(loader.normals[vertex.normalIndex].y);
                vertices.push_back(loader.normals[vertex.normalIndex].z);
            }

            // TODO: use correct color
            if(colorEnabled) {
                vertices.push_back(loader.normals[vertex.normalIndex].x);
                vertices.push_back(loader.normals[vertex.normalIndex].y);
                vertices.push_back(loader.normals[vertex.normalIndex].z);
            }

            if(uvEnabled) {
                vertices.push_back(loader.uvs[vertex.uvIndex].x);
                vertices.push_back(loader.uvs[vertex.uvIndex].y);
            }

            indices.push_back(indices.size());
        }
    }

    init(vertices, indices, features);
}

void Object::init(
    const std::vector<float> vertices,
    const std::vector<unsigned int> indices,
    const std::vector<VertexFeature> features
) {
    this->rotation = glm::mat4(1.0f);
    this->scaling = glm::mat4(1.0f);
    this->translation = glm::mat4(1.0f);

    glGenVertexArrays(1, &VAO);
    glBindVertexArray(VAO);

    glGenBuffers(1, &VBO);
    glBindBuffer(GL_ARRAY_BUFFER, VBO);
    glBufferData(GL_ARRAY_BUFFER, vertices.size() * sizeof(float), vertices.data(), GL_STATIC_DRAW);

    unsigned short stride = 0;
    for(VertexFeature feature : features) {
        stride += getVertexFeatureSize(feature);
    }

    unsigned short initialOffset = 0;
    for(VertexFeature feature : features) {
        unsigned short size = getVertexFeatureSize(feature);
        glEnableVertexAttribArray((int) feature);
        glVertexAttribPointer((int) feature, size, GL_FLOAT, GL_FALSE, stride * sizeof(float), (void*)(initialOffset * sizeof(float)));
        initialOffset += size;
    }

    indexCount = indices.size();
    glGenBuffers(1, &IBO);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, IBO);
    glBufferData(GL_ELEMENT_ARRAY_BUFFER, indexCount * sizeof(unsigned int), indices.data(), GL_STATIC_DRAW);

    textureEnabled = std::find(features.begin(), features.end(), VertexFeature::UV) != features.end();
}

void Object::setTexture(const Texture& texture) {
    this->texture = &texture;
}

void Object::setShader(const ShaderProgram& shader) {
    this->shader = &shader;
}

void Object::setRotation(float angle, glm::vec3 axis) {
    rotation = glm::rotate(glm::mat4(1.0f), angle, axis);
}

void Object::rotate(float angle, glm::vec3 axis) {
    rotation = glm::rotate(rotation, angle, axis);
}

void Object::setTranslation(glm::vec3 translation) {
    this->translation = glm::translate(glm::mat4(1.0f), translation);
}

void Object::translate(glm::vec3 translation) {
    this->translation = glm::translate(this->translation, translation);
}

void Object::scale(float factor) {
    this->scaling = glm::scale(this->scaling, glm::vec3(factor));

    // bug source
    // normalize w component
    for(int i = 0; i < 3; i++) {
        scaling[i][3] /= factor;
    }
}

void Object::render(Camera& camera) {
    glBindBuffer(GL_ARRAY_BUFFER, VBO);
    glBindVertexArray(VAO);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, IBO);

    if(textureEnabled) {
        texture->bind();
        shader->setTexture("tex0", *texture);
    }

    shader->use();

    shader->setMatrix4("camMatrix", camera.getMatrix());
    shader->setMatrix4("modelview", translation * rotation * scaling);

    glDrawElements(GL_TRIANGLES, indexCount, GL_UNSIGNED_INT, nullptr);

    glBindVertexArray(0);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
    glBindBuffer(GL_ARRAY_BUFFER, 0);
}

void Object::free() {
    glDeleteVertexArrays(1, &VAO);
    glDeleteBuffers(1, &VBO);
    glDeleteBuffers(1, &IBO);
}

File: src/object.rs
extern crate gl;
extern crate nalgebra as na;
extern crate nalgebra_glm as glm;

use glm::{Mat4, Vec3};
use gl::types::{GLuint, GLfloat, GLsizeiptr, GLvoid};
use crate::render::Render;
use crate::shader_program::ShaderProgram;
use crate::texture::Texture;
use crate::camera::Camera;
use crate::vertex_feature::VertexFeature;
use crate::object_template::ObjectTemplate;

pub struct Object {
    scaling: Mat4,
    rotation: Mat4,
    translation: Mat4,
    texture_enabled: bool,
    texture: Option<Texture>,
    shader: Option<ShaderProgram>,
    index_count: usize,
    vao: GLuint,
    vbo: GLuint,
    ibo: GLuint,
}

impl Object {
    pub fn new(vertices: Vec<f32>, indices: Vec<u32>, features: Vec<VertexFeature>) -> Self {
        let mut object = Object {
            scaling: Mat4::identity(),
            rotation: Mat4::identity(),
            translation: Mat4::identity(),
            texture_enabled: false,
            texture: None,
            shader: None,
            index_count: 0,
            vao: 0,
            vbo: 0,
            ibo: 0,
        };
        object.init(vertices, indices, features);
        object
    }

    pub fn from_template(object_template: &ObjectTemplate) -> Self {
        Self::new(
            object_template.vertices.clone(),
            object_template.indices.clone(),
            object_template.features.clone()
        )
    }

    fn init(&mut self, vertices: Vec<f32>, indices: Vec<u32>, features: Vec<VertexFeature>) {
        unsafe {
            gl::GenVertexArrays(1, &mut self.vao);
            gl::BindVertexArray(self.vao);

            gl::GenBuffers(1, &mut self.vbo);
            gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
            gl::BufferData(gl::ARRAY_BUFFER, (vertices.len() * std::mem::size_of::<GLfloat>()) as GLsizeiptr, vertices.as_ptr() as *const GLvoid, gl::STATIC_DRAW);

            let stride: u16 = features.iter().map(|feature| feature.size()).sum();

            let mut initial_offset = 0;
            for feature in &features {
                let size = feature.size();
                gl::EnableVertexAttribArray(*feature as GLuint);
                gl::VertexAttribPointer(*feature as GLuint, size as i32, gl::FLOAT, gl::FALSE, (stride * std::mem::size_of::<GLfloat>() as u16) as i32, (initial_offset * std::mem::size_of::<GLfloat>() as u16) as *const GLvoid);
                initial_offset += size;
            }

            self.index_count = indices.len();
            gl::GenBuffers(1, &mut self.ibo);
            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, self.ibo);
            gl::BufferData(gl::ELEMENT_ARRAY_BUFFER, (self.index_count * std::mem::size_of::<u32>()) as GLsizeiptr, indices.as_ptr() as *const GLvoid, gl::STATIC_DRAW);

            self.texture_enabled = features.contains(&VertexFeature::UV);
        }
    }
}

impl Render for Object {
    fn set_texture(&mut self, texture: Texture) {
        self.texture = Some(texture);
    }

    fn set_shader(&mut self, shader: ShaderProgram) {
        self.shader = Some(shader);
    }

    fn set_rotation(&mut self, angle: f32, axis: Vec3) {
        self.rotation = glm::rotate(&Mat4::identity(), angle, &axis);
    }

    fn rotate(&mut self, angle: f32, axis: Vec3) {
        self.rotation = glm::rotate(&self.rotation, angle, &axis);
    }

    fn set_translation(&mut self, translation: Vec3) {
        self.translation = glm::translate(&Mat4::identity(), &translation);
    }

    fn translate(&mut self, translation: Vec3) {
        self.translation = glm::translate(&self.translation, &translation);
    }

    fn scale(&mut self, factor: f32) {
        self.scaling = glm::scale(&self.scaling, &Vec3::new(factor, factor, factor));
    }

    fn render(&self, camera: &Camera) {
        unsafe {
            gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
            gl::BindVertexArray(self.vao);
            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, self.ibo);

            if let Some(shader) = &self.shader {
                shader.use_program();

                if self.texture_enabled {
                    if let Some(texture) = &self.texture {
                        texture.bind();
                        shader.set_texture("tex0", texture);
                    }
                }

                shader.set_matrix4("camMatrix", &camera.get_matrix());
                shader.set_matrix4("modelview", &(self.translation * self.rotation * self.scaling));

                gl::DrawElements(gl::TRIANGLES, self.index_count as i32, gl::UNSIGNED_INT, std::ptr::null());
            }

            gl::BindVertexArray(0);
            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, 0);
            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
        }
    }
}

impl Drop for Object {
    fn drop(&mut self) {
        unsafe {
            gl::DeleteVertexArrays(1, &mut self.vao);
            gl::DeleteBuffers(1, &mut self.vbo);
            gl::DeleteBuffers(1, &mut self.ibo);
        }
    }
}